### DISCUSSION TITLE
Migrating to USWDS 3.0

### DISCUSSION THREAD

**Discussion by thisisdano** [uswds/uswds#4660] [General]
💎 Now that USWDS 3.0 is out, let's talk about migration. We have [a new migration guide](https://designsystem.digital.gov/documentation/migration/) up on our website, and it would be nice to use this space to hear how it's going for you:

- Have you attempted a migration?
- Any bugs or pain points?
- Any feedback on the migration docs?
- Did you try to use packages? Did you see a reduction in your CSS size?
- Did you try to optimize for compile time? How does it compare to your 2.0 compile time?

Whatever your comments or feedback, we like to hear about migration experience (successful and not successful) here in this discussion.

[Check out our 3.0 migration guide](https://designsystem.digital.gov/documentation/migration/) → 
[Read our introduction to USWDS 3.0](https://designsystem.digital.gov/whats-new/updates/2022/04/28/introducing-uswds-3-0/) →

**Comment by thisisdano** [comment]
Hey, so I migrated our Public Sans website to USWDS 3.0 last Friday. It ended up being a pretty good experience, with a good outcome, but there were a couple quirks I'll outline here. There are a few things I found that I'll be turning into issues soon, and I can link them up when I do.

So, in the this report I'll move through the sections of [the USWDS 3.0 migration doc](https://designsystem.digital.gov/documentation/migration/) and talk about what I found, what I did, where I found problems, and how my compile time and CSS size changed as I went through the migration process.

First, the top line:

**Site:** [public-sans.digital.gov](https://public-sans.digital.gov) → 
**Repo:** [github.com/uswds/public-sans](https://github.com/uswds/public-sans) → 

Item | Start | Finish | Change
--- | --- | --- | --- 
**USWDS version** | 2.12.2 | 3.0.0 | — |
**Settings version** | 2.10.0? | 3.0.0 | —
**Compiler setup** | Adapted from USWDS Gulp | USWDS Compile | —
**Compiler packages** | `gulp-sass@4.1.0`, `sass@1.32.8` | `@uswds/compile@1.0.0-beta.2`
**CSS size** | 416 KB | 209 KB | -50%
**Compile time** | 5.3s | 3s | -40%
**Vulnerability audit** | `13` moderate, `17` high, `1` critical | `5` moderate, `7` high | `-8` moderate `-10` high `-1` critical

**Total time for migration:** A little under 2 hours

## 1. Check your current USWDS code and settings versions

Seemed straightforward, but I wasn't 100% about the settings version. They said `2.10.0` but I didn't bother going into GitHub to see if they'd been updated since then. I just proceeded as though they were from `2.10.0` and it worked out OK. The migration guide didn't go all the way back to `2.10.0` but it seemed to work out fine anyway.

## 2. Install the USWDS 3.0 package

No problems here.

## 3. Update your Sass compiler settings and recompile CSS

This is where things started to go sideways a bit. 

### Updating Sass packages 

It was easy enough to follow the instructions and I updated the load paths and the USWDS variables and tasks OK in my gulpfile. I did wonder if the `node_modules` refs in the updated `require()`s needed to have `./` in front of them. It was inconsistent in the docs. It turned out not to matter for me, I left them inconsistent and it worked, but we should fix the inconsistency.

But when I compiled at the end of it all, I got an error:

```
Error: Undefined function.
   ╷
15 │     $grid-to-rem: math.div(
   │ ┌─────────────────^
16 │ │       (grid-base.$system-spacing-grid-base * $unit),
17 │ │       root.$root-font-size-equiv
18 │ │     ) * 1rem;
   │ └─────^
   ╵
```

Turns out not every version of the Sass compiler can handle Sass modules equally well. I had to update my `gulp-sass` package from `4.1.0` to `5.1.0`. This helped a bit, but once I updated I got a new error:

```
sass.sync is not a function
```

Somewhere between my old `gulp-sass` and the new one, it retired `sass.sync`, so I removed that `.sync` from the `sass` function I was using in my gulpfile.

That got me a bit farther, but then I got another error:

```
gulp-sass no longer has a default Sass compiler; please set one yourself.
Both the "sass" and "node-sass" packages are permitted.
For example, in your gulpfile:
const sass = require('gulp-sass')(require('sass'));
```

I updated the `const sass` in my gulpfile to include the line suggested in the error. This compiled!

So, if you're using older versions of `gulp-sass` and `sass` in your gulpfile:

1. Update those packages.
1. Replace any instance of `sass.sync()` with just `sass()`
1. Update your `const sass` to `const sass = require('gulp-sass')(require('sass'));`

This got me compiling but it was pretty slow: 17 seconds.

### Checking settings

And when I checked my site, it looked very bad. Some styles were definitely not loading and the colors were all messed up. What happened was that the `text-` utilities weren't loading and my markup needed them. 

As it turned out, my `$color-settings` utility setting was set to `output: false`. My 2.x site had `$output-all-utilities: true` which was keeping it outputted, but  `$output-all-utilities` is deprecated in 3.0.0, so the utility did not output. I set it back to `output: true` and all was well.

## 4. Integrate any recent USWDS changes

I thought this part might be tedious, but I was surprised that it was pretty straightforward.

I noted that in the migration guide, in the **Older than 2.12.0** section, `$theme-color-success-darker`  should show the OLD value as `60` not `80`. We'll fix that. 

In the guidance, for changed defaults, we recommend checking the settings to see if the project uses the old default. If we use the old default, delete the setting from our settings file so it uses the system default. 

OK, but what if I use the new default? Just leave it? Probably delete the setting if it uses _either_ the old default or the new default. And check to see if you use that setting in your codebase as well.

Then, in **Older than 2.11.2** we replace instances of `$theme-site-max-width` with `$theme-grid-container-max-width`. This broke my site. As it turned out, I had an incorrect value set for `$theme-site-max-width` — since USWDS was ignoring it, the site looked OK. But once I updated the setting, USWDS loaded it and the site broke. It turns out I Just wanted to delete that settings variable so it used the USWDS default. Once I deleted the `$theme-grid-container-max-width` and reverted to the default, my site loaded as expected.

**Up to this point, everything had taken about 40 minutes.**

## 5. Update to Sass module syntax

I also expected this to be tedious, but it was not. The process of diffing my settings against the 2.13.3 defaults was — to me! — kind of insightful and I gained some extra appreciation of the specific ways my project differed from USWDS defaults.

Moving from `@import` to `@forward` was simple. 

Adding `@use "uswds-core" as *;` to my custom Sass was simple. 

At the end of the process my files felt much more streamlined.

**This section took about 10 minutes.**

### Using component packages

This is where it started to get tedious. It was very boring looking for `usa-` prefixed components one by one. I ended up just searching for `usa-` and keeping a running list of the components I ran across. 

But the process worked and after I finished I was down to `9` components.

However, my site was totally messed up again.

Turns out I forgot two critical packages: 
- `uswds-utilities`: My site was using lots of utilities, but it wasn't in the component packages list I was using, so I forgot it. Don't forget utilities!
- `uswds-global`: It also turns out that most sites will likely need to `normalize` which is one of the things in `uswds-global`. My site certainly did. If you see strange margins — especially around the outside of your page — you need `uswds-global`.

Once I added those two, all was well again.

**This took about 20 minutes.**
**In the end, it took 13s to build a 289 KB CSS file, a reduction of 127 KB!**


### Managing utility classes

Another cruddy step, since searching for utilities is also quite tedious — then, when you find them, you have to associate them with a utility module, and many modules share the same base class. I ended up just including all the modules associated with a base class for now, so I could see what the new CSS size was, then I'd go from there.

We'll have to work on how to make that process better.

But I did my best — I ended up including 26 utility modules in `$output-these-utilities`: kind of a lot!

**This took about 25 minutes.**
**In the end, it took 9.5s to build a 225 KB CSS file. I got rid of another 65 KB.**

### Update to the sass-embedded compiler

I decided to put this part off for now, and continue on with my CSS-reduction activities. One nice thing about all of this is you kind of get on a roll, and can get pretty excited about shaving a few more KB off your CSS size. So I moved on to checking my breakpoints. 

I recommend that we move this section after **Reduce utility responsive breakpoints** in the docs to stay on task.

### Reduce utility responsive breakpoints

Searching for breakpoints and updating my breakpoint settings was pretty simple. But I do wish that grid breakpoints would be different from utility breakpoints. This could be an additional optimization. 

This only took a couple minutes.

**After this step I was down to 215 KB / 9.5s. Saved another 10 KB.**

Then I decided I would actually try to prune the utilities a bit more, so I dug into just what module I was using for each of the base classes in my code. So, for all the `flex-` utilities, I looked at whether I was using them for **flex-direction** with something like `flex-row` or for **flex-align** with something like `flex-align-start`.

This took about another 20 minutes, but at the end, I was able to shave off another 11 KB.

But noooooo!! My site got all messed up again. Turns out the `margin` utility module actually has three undocumented related modules: `margin` (margin on all sides), `margin-vertical` (margin-top and margin-bottom), and `margin-horizontal` (margin-left and margin-right). 

By only including `margin` in `$use-these-utilities`, my `margin-top` utilities weren't working. Once I added `margin-vertical` to `$use-these-utilities`, everything worked. We need to add these two undocumented modules to the utility module list.

But then everything worked. 

**At the end of this step: 209 KB / 9.1s.**

### Update to the sass-embedded compiler: The return

Everyone should probably use `sass-embedded` with USWDS 3.0. This makes a big difference.

Use `const sass = require("gulp-sass")(require("sass-embedded"));`

But I did this and my compile time didn't change. That was disappointing. However, I still had a `sass.compiler = require("sass");` line in my gulpfile. When I removed that line, that's when I saw a big change.

So:
1. Use `const sass = require("gulp-sass")(require("sass-embedded"));`
1. Uninstall all Sass packages that aren't `gulp-sass` or `sass-embedded`
1. Remove any `sass.compiler = require("sass");` line in your gulpfile.

Woohoo, now it's compiling in under 3s!

**Final before/after: 416 KB / 209 KB (-50%), 5.3s / 3s (-40%)**


## Conclusion

There were a couple small bugs that I've outlined here and a couple places where we can improve the docs. But, it didn't actually take that long, and it was a success in the end. 

It really got me thinking about optimizing for performance, and I'm excited to do more. 

Utilities can take up a LOT of CSS space. I'd like to rewrite some of my markup to use fewer utilities, and I bet I can get significantly under 200 KB — maybe even down to about 100 KB or so. 

In this process, I also migrated from a custom USWDS Gulp compile workflow to USWDS Compile, but that's another story! It's a little less easy.

**Reply by ghost** [reply]
It is an wonderful migration guide! Thank you!

I am new with uswds theme. Our website uses uswds as our base theme, recently we noticed that we cannot implement some components to our site after upgrading Drupal 8 to 9. I just checked our site uses uswds 1.4.2, probably it causes the issues. 

I need to migrate it to the latest version 3.0. Is it possible for me to jump it from 1.4.2 to 3.0? What all necessary changes do we need in our codes? Do you have any guide for migrating version 1 to 3?

Thanks again!

**Reply by mejiaj** [reply]
@xiaohuima-2020 responded in your [discussion](https://github.com/uswds/uswds/discussions/4895#discussioncomment-3318185).

**Comment by mbucknell** [comment]
I just wanted to commend you on what I think is an excellent migration guide particularly for the optimization section. The build for our projects is custom and doesn't use gulp but I was easily able to figure out what to change. But I also want to encourage others who build their style sheets using sass to try out using component packages (it's step 6 in the migration guide). For our monitoring location page, I reduced the unzipped asset from 1.4MB to 211KB. For my current project that we have not yet made public, we went from 433KB to 183KB. Yes it is a little tedious to go through markup (and in my case javascript since we render much of the page dynamically) to figure out what you need to add but it is worth it!

In my current project which uses Vue Single File Components we had avoided putting style within the components because we couldn't find a good way to do this without a lot of bloat in the asset size. We will soon be exploring whether we can put the Sass style in the Vue SFC components while still using the USWDS mixins without generating huge style sheets. I'll let you all know how that goes when we get to it.
